zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs

```diff
@@ -201,20 +201,13 @@ fn checked_add_transaction_weighted_random(
 /// If some transactions have not yet been chosen, returns the weighted index and the transaction.
 /// Otherwise, just returns the transaction.
 fn choose_transaction_weighted_random(
-    candidate_txs: &[VerifiedUnminedTx],
-    mut weighted_index: WeightedIndex<f32>,
+    candidate_txs: &mut Vec<VerifiedUnminedTx>,
+    weighted_index: WeightedIndex<f32>,
 ) -> (Option<WeightedIndex<f32>>, VerifiedUnminedTx) {
     let candidate_position = weighted_index.sample(&mut thread_rng());
-    let candidate_tx = candidate_txs[candidate_position].clone();
+    let candidate_tx = candidate_txs.swap_remove(candidate_position);
 
     // Only pick each transaction once, by setting picked transaction weights to zero
-    if weighted_index
-        .update_weights(&[(candidate_position, &0.0)])
```

Comment:
Optional: Once we're sure this works, did you want to submit a PR to update the `update_weights` documentation?
https://github.com/rust-random/rand/blob/master/src/distributions/weighted_index.rs#L144

Reply:
Yeah that sounds good, I'll open a PR to fix the bug rather than update the documentation because I think this is just undesirable behaviour. Perhaps by calculating `total_weight` by summing up all of the weights rather than subtracting the old weight when the new weight is zero and the distribution's total weight is less than twice the old candidate weight.

Reply:
I don't think it's possible to fix the bug, because doing an efficient weights update requires re-using partial sums. That works fine for integers, but floating-point sums can't be re-ordered.